Commands rework

itoko opened the pull request:
This PR address the following issues in current class design around commands:

- A command takes only the single channel in its arguments (We need to handle multiple channels especially in Acquire command)
- How to create "Command with channel context" and handle it as a Schedule.

In this PR, we make `Command` be a creator of "Command with channel context" (named `Instruction`).
Using these `Command`s, we can make Schedule like as below (see `test/python/pulse/test_schedule.py` for a working example code).
```
gp0 = gaussian(duration=20, name='pulse0', amp=0.7, t0=9.5, sig=3)
gp1 = gaussian(duration=20, name='pulse1', amp=0.5, t0=9.5, sig=3)
fc_pi_2 = FrameChange(phase=1.57)
acquire = Acquire(10)
sched = Schedule(device)
sched.insert( 0, gp0(device.q[0].drive))
sched.insert(30, gp1(device.q[1].drive))
sched.insert(60, FrameChange(phase=-1.57)(device.q[0].drive))
sched.insert(60, gp0(device.q[0].drive))
sched.insert(90, fc_pi_2(device.q[0].drive))
sched.insert(90, acquire(device.q[0], device.mem[0], device.c[0]))
sched.insert(90, acquire(device.q[1], device.mem[1], device.c[1]))
```
Also I'd like to make `Schedule` immutable. It will be addressed in the next PR.

itoko committed:
make commands callable

itoko committed:
make channels hashable

itoko committed:
add working example code of Schedule

taalexander commented on qiskit/pulse/commands/acquire.py:
```diff
@@ -5,15 +5,62 @@
+from .pulse_command import PulseCommand
+
+
+class AcquirePulse(Pulse):
```
I take issue with the naming. An acquire is not a pulse, nor for example, is a phase change. This nomenclature may be confusing to users.

itoko replied:
I changed the name of the interface for "commands with channel context" from `Pulse` -> `Instruction`, e.g. 
```
class AcquireInstruction(Instruction):
```
I think it looks better, we are looking for even better names though.

itoko committed:
reorder class defs and add prints

itoko committed:
simplify Schedule

itoko committed:
rename PulseChannel->Channel

itoko committed:
merge master change

itoko committed:
lint

itoko committed:
add getters for qobj creation

itoko committed:
simplify snapshot command

taalexander commented on qiskit/pulse/commands/acquire.py:
```diff
@@ -68,26 +70,81 @@ def __eq__(self, other):
+        return AcquirePulse(self, qubits, mem_slots, reg_slots)
+
+
+class AcquirePulse(Pulse):
```
I think there is a lot of reusable logic here that could be done once if this just inherited from a `CommandSchedule`, furthermore an acquire is not a `Pulse`.

itoko replied:
This will be addressed in another `schedule-rework` PR.

taalexander commented on qiskit/pulse/commands/frame_change.py:
```diff
@@ -41,3 +45,43 @@ def __eq__(self, other):
+        return FrameChangePulse(self, channel)
+
+
+class FrameChangePulse(Pulse):
```
Framechange is also not a pulse, this is a `Schedule` at this point.

itoko replied:
I changed the name of the interface for "commands with channel context" from Pulse -> Instruction, e.g.
```
class FrameChangeInstruction(Instruction):
```
And `Schedule` has the same `Instruction` interface.
```
class Schedule(Instruction, ScheduleNode):
```
Can we push making `Instruction` be a `Schedule` to `functional-style` PR? 
(since I currently have no clear implementation image to do so.)
In this PR, I'd like to focus on making `Command` callable to show how UI around commands changes to internal users earlier.

taalexander reviewed (changes requested):
This is looking on the right track, I have just quickly reviewed today not a full review.

itoko committed:
rename Pulse -> Instruction

itoko commented:
@taalexander Currently, I implemented `__call__` of Commands, but what if we switch it to `__rshift__`? (inspired by your syntax sugars)
```
sched.insert( 0, gp0 >> device.q[0].drive)
sched.insert(30, gp1 >> device.q[1].drive)
sched.insert(60, FrameChange(phase=-1.57) >> device.q[0].drive)
sched.insert(90, fc_pi_2 >> device.q[0].drive)
sched.insert(90, acquire >> device.q[0], device.mem[0], device.c[0])
```
This looks cooler for me. What do you think?

itoko committed:
Merge branch 'pulse_channel' into commands-rework

itoko committed:
simplify Instruction

taalexander commented:
I think it is good about how else we can use dunder methods to make it cleaner to right a pulse schedule. What I worry about the use of `>>` is that I think it would often be confused as the equivalent of `<< - shift_time`. Also, Python already expose a concise way of calling a function to return an object which we already use (`__call__`).

itoko committed:
refactor schedule-related interface

itoko committed:
unify exceptions into PulseError

itoko committed:
add a test

itoko commented:
OK. I'll leave it as `__call__` in this PR.

Commit:
update assembler

Commit:
update lo frequency

chriseclectic committed:
fix numpy casting error with *= (#2062)

nonhermitian committed:
Allow for list of ints as layout (#2058)

* int list as layout

* make integer layout a method of Layout

* integer_layout -> intlist_layout

* similar interface to generate_trivial_layout

* move the test to test_compile

* test for the errors

* changelog

* If the intlist is longer that your quantum register, map them to None.

* valid names

* the name should make it obvious that constructs an object

* lint

itoko committed:
merge command name rework

itoko committed:
add check length of mem_slots

itoko committed:
Merge branch 'commands-rework' into commands-rework-assembler

diego-plan9 committed:
Remove underscore from quantum_info, providers (#2060)

* Remove underscore from basicaertools

* Remove underscore from quantum_info

nonhermitian committed:
cast to uint32 (#2063)

nonhermitian committed:
Allow for bit string distance measures (Hamming dist) in plot_histogram (#2064)

* allow for sorting histograms by hamming dist

* lint

* Update qiskit/tools/visualization/_counts_visualization.py

Co-Authored-By: nonhermitian <nonhermitian@gmail.com>

* add to changelog

* fix spelling

ajavadia committed:
Remove qargs/cargs/circuit from instructions, in favor of circuit keeping context (#1816)

* remove qarg and carg from Instruction, modify circuit._attach

* update circuit_to_dag, remove qarg/carg from all gates

* fixing decomposition rule definitions

* update to decomposition rules

* bring measure qasm under QuantumCircuit

* add args context to InstructionSet

* fix extension tests, update matrix unitary gate to no qargs

* update snapshot instruction

* fix barrier insertion pass

* update transpiler passes for new format, rest of decomp rules

* add instruction shapes (num_qubits, num_clbits)

add instruction shapes (num_qubits, num_clbits)

* fix tests except initialize

* initialize is a generic instruction with decomposition rules

* move decompositions to generic instructions, add circuit_to_instruction

* add circuit.append()

* generalize dag_to_circuit

* use copy in circuit_to_dag and dag_to_circuit, temporary slowdown

* allow unroller pass to unroll generic instructions, not just gates

* initialize as Instruction that recursively decomposes

* revert to only gates being unrolled, and method for circuit_to_gate for now

* make qargs and cargs optional in circuit.append and dag.apply_operation_back

* add circuit.to_gate() method to wrap the convertor

* cleanup initialize

* use a list of Gate for a Gate.definition, decompose based on that

* add flag to indicate instruction reversibility

* decompose pass does not error on fundamental gates

* rename init to initialize in qobj

* circuit.decompose() method

* remove instruction._modifiers method

* remove instruction.circuit property

* CHANGELOG

* allow general subroutines via instruction definitions and update unroller

* instruction.reverse() -> instruction.mirror()

* inplace option for Instruction mirror and inverse

* base gates should not modify with inverse. add tests

* flatten register in circuit_to_instruction conversion

jwoehr committed:
Add quiet and output to job monitor to suppress or redirect status messages (#2037)

* job monitor takes quiet= (bool:default False) and to_file = (file:default None == sys.stdout)

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* broke too-long lines that failed pylint

chriseclectic committed:
fix sdg definition (#2069)

maddy-tod committed:
Fixed mpl drawer output (#2061)

* Updated the mpl drawer so that the output is no longer messy

* Renamed gate_index to gate_anchor

* Fixed off by 1 error - gates now begin in column 0 not 1

* Fixed linting

taalexander reviewed (approved):
This is looking good. There are some things I would change, but these can be done in later PR. We need a lot more tests, of pretty much every component. After tests we should finalize this and submit PR to terra master. Then we can make future PRs with respect to master.

nonhermitian committed:
Return minimal bandwidth best layout (#2021)

* Compute a minimal bandwidth best layout

* flip cx gates in offending test

* flipped wrong test

* update changelog

* fixed tests

itoko committed:
Merge pull request #18 from nkanazawa1989/commands-rework-assembler

[WIP] Commands rework assembler

ajavadia committed:
test unrolling for all gates (#2072)

taalexander commented:
I was under the impression that we would submit the assembler as a separate PR to master? Could we break these into two? With this branch being submitted and then the assembler going as a separate PR based on this branch? This will make it easier to review

mtreinish committed:
Just use a list not split a string (#2071)

In the matplotlib circuit drawer backend there was a statically defined
list of gates to treat as wide. For some reason this list was defined by
build a space separated string of the gates and then running the split()
method on that string. This is unecessary and we can just define a list
of these gates. This commit does just that and defines those gates as a
list and removes the split() uasge.

1ucian0 committed:
remove_op_node is part of the DAG API (#2024)

* remove_op_node is part of the DAG API

* lint

* docstirng

* raises in case of non node.type != op

* test_remove_op_node_longer

* Update qiskit/dagcircuit/dagcircuit.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* new apply_operation_back
api

ewinston committed:
Add arbitrary unitary gate (#1961)

* start unitary class and add some tests

tests passing

insert Unitary in QuantumCircuit

minor commit

add power

linting, bug fix

added tests

remove check_qubit during reapply

Extend InstructionParameter for containers

Extend `InstructionParameter`, adding support for containers (`list`
and `numpy.ndarray`. Add tests for the field.

Resync elapsed time with ISO 80000-3 annex

fixes bug (uses PR #1970)

fix silent bad type

* linting

* typo

* remove matrix method. Add _representation method

* bug fixes

* fix init with Unitary. Overload __eq__.

* minor fix

* adapt to pr #816 removing qargs from instructions.

* minor linting

* remove reapply and decomposition methods

these methods are not longer needed after recent upstream changes.

* remove unused import

chriseclectic committed:
fix occasionally failing tests from non-unique Kraus conversion (#2075)

mtreinish committed:
Migrate visualization to top level namespace (#1878)

* Migrate visualization to top level namespace

This commit moves the visualization code out of tools into it's own top
level namespace, 'qiskit.visualization'. The visualization was a bit out
of place in the tools directory and is really it's own top level thing
at this point. However, because we declared the public visualization
interfaces stable as part of the 0.7 release we must maintain backwards
compatibility. In order to facilitate that all the public APIs are still
accessible via 'qiskit.tools.visualization'. However, private
api/modules are not accessible off of 'qiskit.tools.visualization'
anymore.

* Add changelog

* Fix lint

Sigh pylint, raising an error for both using a wildcard import and not
using all the imports from the wild card.

* Rebase updates

* Move tests to match new dir structure

* Update test_visualization import per review comment

* Remove underscores from visualization modules

* Fix lint

1ucian0 committed:
ResourceEstimation AnalysisPass (#1889)

* count operations pass

* empty dag test

* rename

* test

* test rename

* test name

* Update qiskit/transpiler/passes/resource_estimation.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* count_ops

* count_ops

* lint

* changelog

* depth pass

* size pass

* size pass

* count_ops pass

* size pass

* NumTensorFactor pass

* resource calculation metapass

* lint

* unused import

* changelog

* lint

* Update qiskit/transpiler/passes/__init__.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/width.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/num_tensor_factor.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/depth.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/depth.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/size.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/width.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/size.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/num_tensor_factor.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* renaming

mtreinish committed:
Don't pass an empty RunConfig to assemble_circuits (#2011)

* Don't pass an empty RunConfig to assemble_circuits

This commit modifies the deprecated circuits_to_qobj() converter
function to handle the case where the run_config kwarg is not set
properly. Previously the function was initializing an empty RunConfig
option unecessarily and passing that into assemble_circuits(). However,
this was not necessary because assemble_circuits() can handle the case
where run_config is not set.  So this just removes that and passes None
directly to assemble_circuits() which fixes potential validation errors.

Fixes #1988

* Fix lint

* More lint fixes

Commit:
Merge branch 'master' of github.com:Qiskit/qiskit-terra into commands-rework

itoko commented:
Moved to https://github.com/nkanazawa1989/qiskit-terra/pull/31
